### PR TITLE
Handle lists being modified by use of ~

### PR DESCRIPTION
**Pull request by ChrisJefferson**
Adjust how lists are built when ~ is used, so we do not crash
even if the list is modified as it is being constructed.

This also makes the behaviour of GAP consistent between bytecode
and immediate code.

### CODE REVIEW

File: src/exprs.c [hunk 1]
```diff
@@ -919,9 +919,18 @@ Obj             EvalPermExpr (
 **
 **  'EvalListExpr'  just  calls 'ListExpr1'  and  'ListExpr2' to evaluate the
 **  list expression.
+**
+**  The 'safe' argument to ListExpr1 and ListExpr2 handle code which abuses ~
+**  access and change the partially built list.
```
Comment (fingolfin) [inline src/exprs.c]: handle -> handles ? though I don't see how an argument can "handle" anything by itself?

File: src/exprs.c [hunk 2]
```diff
@@ -919,9 +919,18 @@ Obj             EvalPermExpr (
...
+**
+**  When 'safe' is 1, we use a slower code path which ensures the list is
+**  in a valid state after each element is added, and not crash if the
```
Comment (fingolfin) [inline src/exprs.c]: missing words... perhaps "... is added, and that we don't crash ..."

File: src/exprs.c [hunk 3]
```diff
@@ -999,8 +1008,9 @@ Obj             EvalListTildeExpr (
 **  '[ [1], ~[1] ]'  requires that the value of  one subexpression is entered
 **  into the list value before the next subexpression is evaluated.
 */
-Obj ListExpr1 (
-    Expr                expr )
+ALWAYS_INLINE Obj ListExpr1 (
```
Comment (fingolfin) [inline src/exprs.c]: Why `ALWAYS_INLINE`?

File: src/exprs.c [hunk 4]
```diff
@@ -1015,15 +1025,21 @@ Obj ListExpr1 (
     else {
         list = NEW_PLIST( T_PLIST, len );
     }
-    SET_LEN_PLIST( list, len );
+    if(safe) {
```
Comment (fingolfin) [inline src/exprs.c]: add space after `if`

File: src/exprs.c [hunk 5]
```diff
@@ -1053,17 +1069,25 @@ void ListExpr2 (
           {
             if (posshole == 1)
               {
-                SET_FILT_LIST(list, FN_IS_NDENSE);
+                if(!safe || TNUM_OBJ(list) == T_PLIST) {
```
Comment (fingolfin) [inline src/exprs.c]: add space after `if` (or just `git clang-format`)

File: tst/testinstall/tilde.tst [hunk 6]
```diff
@@ -85,4 +85,15 @@ true
 gap> i := InputTextString( "local a; a := 10; return rec(a := a*10, b := ~);" );;
 gap> r := rec(a := ~, b := ReadAsFunction(i)(), c := ~.b.a);
 rec( a := ~, b := rec( a := 100, b := ~.b ), c := 100 )
+gap> [Length(~),Length(~),Length(~)];
+[ 0, 1, 2 ]
+gap> (function() return [Length(~),Length(~),Length(~)]; end)();
+[ 0, 1, 2 ]
+gap> rem := function(l) Remove(l); return 1; end;;
+gap> [2,rem(~),rem(~),rem(~)];
```
Comment (fingolfin) [inline tst/testinstall/tilde.tst]: Wow, that's pretty evil :-)

Comment (ssiccha) [reply tst/testinstall/tilde.tst]: I agree

File: src/exprs.c (hunk 3 above)
Comment (ChrisJefferson) [reply src/exprs.c]: Will fix all other comments.

With this one, ALWAYS_INLINE because I want the 'safe' compiled out when safe=0, so the fast route (without tilde) is as fast as it used to be.

File: tst/testinstall/tilde.tst (hunk 6 above)
Comment (ChrisJefferson) [reply tst/testinstall/tilde.tst]: I don't expect anyone to ever write this, but I feel there shouldn't be ways of making the interpreter crash (at least, without calling ALL_CAPS functions).

Comment (fingolfin) [reply tst/testinstall/tilde.tst]: Yeah, sure, it's good to add that test. People do evil things, we shouldn't reward them for it with a crash ;-)

File: src/exprs.c (hunk 3 above)
Comment (fingolfin) [reply src/exprs.c]: ok

**Comment by ChrisJefferson** [comment]
Hopefully changes made, added to release notes.

File: src/exprs.c [hunk 7]
```diff
@@ -919,9 +919,19 @@ Obj             EvalPermExpr (
 **
 **  'EvalListExpr'  just  calls 'ListExpr1'  and  'ListExpr2' to evaluate the
 **  list expression.
+**
+**  The 'safe' argument to ListExpr1 and ListExpr2 should be set to 1 when
+**  the list expression may be modified when its elements are being
```
Comment (fingolfin) [inline src/exprs.c]: "when its elements are being constructed" -> "while its ..." ? (It took me a moment to parse the current phrasing, but then I did understand it, so it's probably fine as it is -- I just try to be super sensitive and try to simulate how I might read a paragraph if I knew less about the kernel; but of course that's imperfect. Perhaps e.g. @ssiccha would be willing to chime and say whether the above text seems clear to him -- hopefull it is, but  if not, it'd be useful to know which bits are not.

File: src/exprs.c [hunk 8]
```diff
@@ -919,9 +919,19 @@ Obj             EvalPermExpr (
 **
 **  'EvalListExpr'  just  calls 'ListExpr1'  and  'ListExpr2' to evaluate the
 **  list expression.
+**
+**  The 'safe' argument to ListExpr1 and ListExpr2 should be set to 1 when
```
Comment (fingolfin) [inline src/exprs.c]: Shouldn't this text be part of the documentation comment of `ListExpr1` and `ListExpr2` instead?

I understand that there are prototypes for those two functions below, but considering the rest of their documentation is elsewhere... Perhaps we could move them up, before this function, to avoid using prototypes? Then they could also be made static.

File: src/exprs.c [hunk 9]
```diff
@@ -1053,18 +1065,25 @@ void ListExpr2 (
           {
             if (posshole == 1)
               {
-                SET_FILT_LIST(list, FN_IS_NDENSE);
+                if (!safe || TNUM_OBJ(list) == T_PLIST) {
```
Comment (fingolfin) [inline src/exprs.c]: Hmm, is that safe? Couldn't the tnum be, say `T_PLIST_DENSE` at this point? That would suggest using `if (!safe || IS_PLIST(list))` - or perhaps even `IS_LIST`?

But why is this necessary at all? `SET_FILT_LIST` uses a tnum dispatch table -- if the user did something really bad here, wouldn't that just trigger an error via that dispatch table?

File: src/exprs.c [hunk 10]
```diff
@@ -919,9 +919,19 @@ Obj             EvalPermExpr (
...
 **  list expression.
+**
+**  The 'safe' argument to ListExpr1 and ListExpr2 should be set to 1 when
+**  the list expression may be modified when its elements are being
+**  constructed. This can occur when an element of the list refers to ~.
+**
+**  When 'safe' is 1, we use a slower code path which ensures the list is
+**  in a valid state after each element is added, and we don't crash if the
+**  size or TNUM of the list is changed during the construction of elements
+**  in the list.
```
Comment (fingolfin) [inline src/exprs.c]: Perhaps that should say "we don't crash if the size or type of the list changed, or even if gets swapped for a non-list object" ? After all, the user could use SWAP_OBJ to replace it by, say, a bigint.

OTOH, perhaps I am just overthinking that; and people using SWAP_OBJ deserve what they get ;-).

File: src/exprs.c [hunk 11]
```diff
@@ -919,9 +919,19 @@ Obj             EvalPermExpr (
 **
 **  'EvalListExpr'  just  calls 'ListExpr1'  and  'ListExpr2' to evaluate the
 **  list expression.
+**
+**  The 'safe' argument to ListExpr1 and ListExpr2 should be set to 1 when
+**  the list expression may be modified when its elements are being
+**  constructed. This can occur when an element of the list refers to ~.
+**
+**  When 'safe' is 1, we use a slower code path which ensures the list is
```
Comment (fingolfin) [inline src/exprs.c]: Hmm, when I was reading the code below, I constantly had to refer back to this comment, because I could not remember whether `safe` meant: either "we are safe; it is safe to assume that the list won't be messed with", or perhaps "play it safe; use safe code which works even if the list gets messed with".

Since these are opposite, that's a bit of a problem, IMHO. Is it just me? Perhaps we can find a better name. Say `protectAgainstListModification` or `safelyGrowList`, `safelyExtendList`; or perhaps reverse the value, and say `assumeListIsPlist` or ...? Sorry, no great idea.

File: src/exprs.c (hunk 9 above)
Comment (ChrisJefferson) [reply src/exprs.c]: You are right, let me rethink this.

File: src/exprs.c (hunk 11 above)
Comment (ChrisJefferson) [reply src/exprs.c]: I could just call it `tildeInUse`, seeing as that's all we use it for :)

Comment (fingolfin) [reply src/exprs.c]: Yes, that actually sounds very good to me!

File: src/exprs.c (hunk 7 above)
Comment (ssiccha) [reply src/exprs.c]: @fingolfin @ChrisJefferson Apart from knowing what exactly an expression is (`system.h` says its a `UInt`, but that's all I found out quickly), the formulation is clear to me.

Comment (ssiccha) [reply src/exprs.c]: > "when its elements are being constructed" -> "while its ..." 
"while its" also sounds better to me

**Comment by ChrisJefferson** [comment]
Now changed/improved

File: src/exprs.c [hunk 12]
```diff
@@ -998,9 +998,15 @@ Obj             EvalListTildeExpr (
...
+**  'tildeInUse' is 1 when this list is part of the value of ~.
+**
+**  When 'tildeInUse' is 1, use a slower code path which ensures the list is
+**  in a valid state after each element is added and handles the elements
+**  and TNUM of the list changing whenever a new element is constructed.
+**
 */
-Obj ListExpr1 (
-    Expr                expr )
+ALWAYS_INLINE Obj ListExpr1(Expr expr, Int safe)
```
Comment (fingolfin) [inline src/exprs.c]: The documentation now says `tildeInUse`, great -- but the code still uses `safe` :-)

File: src/exprs.c [hunk 13]
```diff
@@ -998,9 +998,15 @@ Obj             EvalListTildeExpr (
 **  This two step allocation  is necessary, because  list expressions such as
 **  '[ [1], ~[1] ]'  requires that the value of  one subexpression is entered
 **  into the list value before the next subexpression is evaluated.
+**
+**  'tildeInUse' is 1 when this list is part of the value of ~.
```
Comment (fingolfin) [inline src/exprs.c]: OK. Could you please adjust lines 984/985? You could copy&past this over them
```C
*F  ListExpr1(<expr>,<tildeInUse>) . . . .  make a list for a list expression
*F  ListExpr2(<list>,<expr>,<tildeInUse>) . . enter sublists for a list expr.
```

File: src/exprs.c [hunk 14]
```diff
@@ -1015,15 +1021,18 @@ Obj ListExpr1 (
     else {
         list = NEW_PLIST( T_PLIST, len );
     }
-    SET_LEN_PLIST( list, len );
+    if (safe) {
+        SET_LEN_PLIST(list, 0);
+    }
+    else {
+        SET_LEN_PLIST(list, len);
+    }
```
Comment (fingolfin) [inline src/exprs.c]: Could also use `SET_LEN_PLIST(list, tildeInUse ? 0 : len);` for compactness -- but of course it's also perfectly fine as it is :-).

File: src/exprs.c [hunk 15]
```diff
@@ -1053,18 +1062,25 @@ void ListExpr2 (
...
+            }
             CHANGED_BAG( list );
           }
 
     }
-    if (!posshole)
-      SET_FILT_LIST(list, FN_IS_DENSE);
-
+    if (!posshole && !safe) {
+        SET_FILT_LIST(list, FN_IS_DENSE);
```
Comment (fingolfin) [inline src/exprs.c]: Why do we need the `&& !safe` (resp. `&& !tildeInUse`) here? Isn't the list fully constructed at this point?

Ahh, because we might have used unbind while constructing it, through the use of `~`? Is that the reason?
Perhaps a small comment could be added to point that out (or, if I am wrong, then point out the actual reason)?

File: src/exprs.c (hunk 12 above)
Comment (ChrisJefferson) [reply src/exprs.c]: bla!

File: src/exprs.c (hunk 15 above)
Comment (ChrisJefferson) [reply src/exprs.c]: I'm just assuming that SET_FILT_LIST is always an optimisation, and as you say, who knows what state the list is in, or if it's even a list.
